Content.Server/Fluids/EntitySystems/MoppingSystem.cs

```diff
@@ -0,0 +1,350 @@
+        SubscribeLocalEvent<TransferCompleteEvent>(OnTransferComplete);
+    }
+
+    private void OnAfterInteract(EntityUid uid, AbsorbentComponent component, AfterInteractEvent args)
+    {
+        if (!args.CanReach) // if user cannot reach the target
+        {
+            return;
+        }
+
+        if (args.Handled) // if the event was already handled
+        {
+            return;
+        }
+
+        _solutionSystem.TryGetSolution(args.Used, AbsorbentComponent.SolutionName, out var absorbedSolution);
+
+        if (absorbedSolution is null)
+        {
+            return;
+        }
+
+        var toolAvailableVolume = absorbedSolution.AvailableVolume;
+        var toolCurrentVolume = absorbedSolution.CurrentVolume;
+
+        // For adding liquid to an empty floor tile
+        if (args.Target is null) // if a tile is clicked
+        {
+            ReleaseToFloor(args.ClickLocation, component, absorbedSolution);
+            args.Handled = true;
+            args.User.PopupMessage(args.User, Loc.GetString("mopping-system-release-to-floor"));
+            return;
+        }
+        else if (args.Target is not null)
+        {
+            // Handle our do_after logic
+            HandleDoAfter(args.User, args.Used, args.Target.Value, component, toolCurrentVolume, toolAvailableVolume);
+        }
+
+        args.Handled = true;
+        return;
+    }
+
+    private void ReleaseToFloor(EntityCoordinates clickLocation, AbsorbentComponent absorbent, Solution? absorbedSolution)
+    {
+        if ((_mapManager.TryGetGrid(clickLocation.GetGridId(EntityManager), out var mapGrid)) // needs valid grid
+            && absorbedSolution is not null) // needs a solution to place on the tile
+        {
+            TileRef tile = mapGrid.GetTileRef(clickLocation);
+
+            // Drop some of the absorbed liquid onto the ground
+            var releaseAmount = FixedPoint2.Min(absorbent.ResidueAmount, absorbedSolution.CurrentVolume); // The release amount specified on the absorbent component, or the amount currently absorbed (whichever is less).
+            var releasedSolution = _solutionSystem.SplitSolution(absorbent.Owner, absorbedSolution, releaseAmount); // Remove releaseAmount of solution from the absorbent component
+            _spillableSystem.SpillAt(tile, releasedSolution, "PuddleSmear");                                        // And spill it onto the tile.
+        }
+    }
+
+    // Handles logic for our different types of valid target.
+    // Checks for conditions that would prevent a doAfter from starting.
+    private void HandleDoAfter(EntityUid user, EntityUid used, EntityUid target, AbsorbentComponent component, FixedPoint2 currentVolume, FixedPoint2 availableVolume)
+    {
+        // Below variables will be set within this function depending on what kind of target was clicked.
+        // They will be passed to the OnTransferComplete if the doAfter succeeds.
+
+        EntityUid donor;
```

Comment:
To Increase readability here:

If you extract the `DoAfterEventArgs` construction and the `WaitDoAfter` call to another function, you can then call `DoMopInteraction(donor, acceptor, donorSolutionName, acceptorSolutionName, transferAmount, delay, msg, sfx)` at the 3 places the interaction is decided. This removes the weird control flow, and every block will return, instead of having the 3 situations where the flow falls through to the bottom of the function.

Reply:
Oh, didn't see this until now. Thank you! I'll give this a shot.

Reply:
Done!